R/SEIRDAge_interventions.R

```diff
@@ -0,0 +1,452 @@
+#' which act to reduce the degree of social contact between individuals. This is is
+#' done by splitting any simulation of the model into 3 intervals and utilising a different
+#' contact matrix in each interval to simulate the introduction of NPIs through
+#' time. This function relies on the
+#' package deSolve to numerically integrate the set of equations above.
+#'
+#' @param object An object of the class SEIRDAge_interventions.
+#' @param times (vector) time sequence over which to solve the model.
+#'        Must be of the form seq(t_start,t_end,by=t_step). Default time series
+#'        is seq(0, 100, by = 1).
+#'
+#' @param solve_method A string indicating the chosen numerical integration
+#' method for solving the ode system. Default is `lsoda` which is also the
+#' default for the ode function in the deSolve package used in this function.
+#'
+#' @return a data frame containing the time vector and time series of S, R, I and
+#' D population fractions for each age group outputs with incidence numbers
+#' for each age group.
+#'
+#' @export
+#' @importFrom magrittr "%>%"
+#' @importFrom rlang .data
+setMethod(
+  "run", 'SEIRDAge_interventions',
+  function(object, times, solve_method = 'lsoda') {
+
+    # error if times is not a vector or list of doubles
+    if(!is.double(times)){
+      stop('Evaluation times of the model storage format must be a vector.')
+    }
+    if (is.null(unlist(object@transmission_parameters)))
+      stop("Transmission parameters must be set before running.")
+    if (is.null(unlist(object@initial_conditions)))
+      stop("Initial conditions must be set before running.")
+    if (is.null(unlist(object@intervention_parameters)))
+      stop("Intervention parameters must be set before running.")
+
+    # fetch number of age catagories
+    age <- object@n_age_categories
+
+    # set initial state vector
+    state <- c(S = object@initial_conditions$S0,
+               E = object@initial_conditions$E0,
+               I = object@initial_conditions$I0,
+               R = object@initial_conditions$R0,
+               D = object@initial_conditions$D0,
+               cc = rep(0, age))
+
+    dt = times[2]-times[1]
+    # set parameters vector
+    parameters <- c(isolated_frac = object@transmission_parameters$isolated_frac,
+                    beta_isolated = object@transmission_parameters$beta_isolated,
+                    beta_not_isolated = object@transmission_parameters$beta_not_isolated,
+                    kappa = object@transmission_parameters$kappa,
+                    gamma = object@transmission_parameters$gamma,
+                    mu = object@transmission_parameters$mu)
+
+    # function for RHS of ode system
+    Intervention = GetIdxIntervention(object, times)
+
+    right_hand_side <- function(t, state, parameters) {
+      with(
+        as.list(c(state, parameters)),
+        {
+
+          intv <- Intervention[which(as.integer(t)==times)]
```

Comment:
I'm not quite sure I understand these functions to do with the intervention. In particular, I don't quite see why we are using "times" (which I think is the times at which we report the solution to the ODE) to determine something on the RHS of the ODE? (Which I assume is the index of the particular intervention in place?) I think we would want a function that accepts "t" as an input and returns the index of the intervention currently happening, but I don't see why that should depend on "times"...perhaps I'm missing something?

Reply:
(Now solved by setting up a function)